Add ProductionCostMultiplier and ProductionTimeMultiplier Traits

MustaphaTR opened the pull request:
I was wrong with that this solves duplicate D2k actors, as prerequisites are still a concern there (still could remove dupe Carryall and Harvester but i don't think it is a good idea to handle some units differently than others.).

Still this solves the duplicate actors on our RAClassic mod for Russia and Turkey and also should solve @SoScared's problem at #15301. Also allows stuff like Generals Tech Oil Refinery and YR Soviet Industrial Plant.

Selling, bounties etc. still use the valued trait for their calculations.

GraionDilach commented:
I don't have a better idea at the moment but this looks too hacky to be approved. Is this really needed for RA Classic? Can't Russia just use a generic queue-based costmultiplier akin to YR Industrial Plant?

MustaphaTR commented:
> Can't Russia just use a generic queue-based costmultiplier akin to YR Industrial Plant?

Our intent is to keep the miscalculated wierd values for Russia so defining them manually for each unit is required.

GraionDilach commented:
Then I vote for introducing a local hack within that mod for it. This is going against OpenRA's evolution itself - exception traits should only be provided for really necessary cases within the primary mods.

I also can't find any documentation about the miscalculation but even if that has a formulae (I presume it just does a wrong rounding) I'm fully behind introducing that formulae as an UseClassicModifierFormulae or something akin to BodyOrientation->UseClassicPredictionFudge than doing it this way.

MustaphaTR commented:
I think the proper way of doing this would to support multipile `Buildable:` traits and making it conditional. A field to override production cost can be the part of `Buildable:`. This would also allow different prereqs per queue and actually solve D2k dupe starport actors. I think that would require a serius refactor tho.

GraionDilach commented:
If we're going that broad, then I'd say splitting Buildable to multiple subtraits, each providing a particular functionality would be the way to go, one trait providing cameo, one the cost, one the prerequisites etc..

Either way, I am still against this PR as-is and I don't know how to resolve the issues anyway at this point.

MustaphaTR commented:
I think i'm getting multipile buildable traits to work but will need some help.

pchote commented:
Needs a rebase now

MustaphaTR committed:
Add ProductionCost/TimeMultiplier

MustaphaTR commented:
Rebased and changed this to be a modifier (with support of making it add additional direct value over the modifier for the stuff needed for RAClassic).

pchote commented:
After our [discussion on IRC](https://logs.openra.net/?year=2018&month=11&day=04#11:54:13) I am concerned about finding a way forward here that won't actively get in the way of solving the other usecases. 

I am overwhelmed with the current glut of PRs introducing design changes across various systems, and don't have any mental bandwidth left to consider the problems here.  I would prefer that we left this on the back burner until things quiet down and there is capacity to properly consider the design of our build-requirements plumbing.

MustaphaTR commented:
I'm not expexting this to be reviewed and merged soon, that's why i seperately PRed this to AS.

Also seems like messed up trying to implement this with interfaces, so i'll take more into this tomorrow and actually make it use interfaces.

pchote commented on OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs:
```diff
@@ -0,0 +1,44 @@
+		public readonly int Multiplier = 100;
+
+		[Desc("After applying the multiplier, also add this to the cost.")]
+		public readonly int ExtraCost = 0;
```
Please remove this as outlined above and simplify the interface return value to `int`.

pchote commented on OpenRA.Mods.Common/Traits/Multipliers/ProductionTimeMultiplier.cs:
```diff
@@ -0,0 +1,44 @@
+		public readonly int Multiplier = 100;
+
+		[Desc("After applying the multiplier, also add this to the build time.")]
+		public readonly int ExtraBuildTime = 0;
```
Please remove this as outlined above and simplify the interface return value to `int`.

pchote reviewed (changes requested):
The code here looks generally fine, and implements the plumbing we need to resolve at least two usecases in the default mods (AI difficulty modifiers and D2K starport prices).

The sticking point is the `ExtraCost` and `ExtraBuildTime` values, which I understand are added to be able to replicate the original RA's arbitrary faction cost modifiers.

This model breaks down as soon as you mix multiple traits. Consider adding an industrial-plant style building to RA classic - the `ExtraCost` won't see the IP modifier, so the cost reduction for Russian players will not be uniformly 25% and players would be quite right to complain that this is a bug. Applying the cost offsets at each step would help, but creates the problem that the final cost depends on the order modifiers are evaluated in.

Arbitrary price/time replacements should be specified as replacement values, and then any real modifiers should act on this replaced value as normal.  Duplicating actor definitions in yaml is far from ideal, but this at least interacts cleanly with modifiers.  I would prefer that we could keep it this way until someone can implement support for multiple `Buildable` traits on an actor.

In the mean time we do have use cases for multiplicative modifiers, and they won't be made redundant by multiple `Buildable` support.  So my preference would be to drop those two fields and then move towards getting the rest of the PR merged rather than letting this sit forever.

This is an important part of the AI difficultly levels that I want to implement for the next release, so I can take this over if you'd prefer not to continue with this.

MustaphaTR commented:
When i implemented the ExtraCost, ExtraBuildTime, i missed a really important point. I can just use multipile traits to do modifications that are more precise than Cost/100 or BuildTime/100. I agree that they can go and i complately forgot i implemented them at the first place.

MustaphaTR commented on OpenRA.Mods.Common/Traits/Multipliers/ProductionCostMultiplier.cs:
```diff
@@ -0,0 +1,44 @@
+		[Desc("Queues that this cost will apply.")]
+		public readonly HashSet<string> Queue = new HashSet<string>();
+
+		Pair<int, int> IProductionCostModifierInfo.GetProductionCostModifier(TechTree techTree, string queue)
```
It is used in the last check, `techTree.HasPrerequisites(Prerequisites)`.

MustaphaTR commented:
Updated.

`D2K starport prices` This PR alone doesn't solve D2k startport prices btw because cost is not the only thing that needs to be changed, prerequisites too. Current implentation would allow us to only remove duplicates for Trike and Carryall which i didn't do in this PR because it would be an inconsistency.

pchote commented:
By starport prices I mean the fluctuating costs, which can be added through a new trait that implements `IProductionCostModifierInfo`.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -160,6 +163,7 @@ void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newO
 			playerPower = newOwner.PlayerActor.TraitOrDefault<PowerManager>();
 			playerResources = newOwner.PlayerActor.Trait<PlayerResources>();
 			developerMode = newOwner.PlayerActor.Trait<DeveloperMode>();
+			techTree = newOwner.PlayerActor.Trait<TechTree>();
```
Please use this cached version for the `.Update()` call below and in `CacheProducibles`.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -434,15 +437,28 @@ public virtual int GetBuildTime(ActorInfo unit, BuildableInfo bi)
+		public virtual int GetProductionCost(ActorInfo unit)
+		{
+			var valued = unit.TraitInfoOrDefault<ValuedInfo>();
+
```
nit: remove newline between valued and null check.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -434,15 +437,28 @@ public virtual int GetBuildTime(ActorInfo unit, BuildableInfo bi)
+				return 0;
+
+			var modifiers = unit.TraitInfos<IProductionCostModifierInfo>().Select(t => t.GetProductionCostModifier(techTree, Info.Type));
+			var cost = Util.ApplyPercentageModifiers(valued.Cost, modifiers);
```
`return Util.ApplyPercentageModifiers(valued.Cost, modifiers);` directly.

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -434,15 +437,28 @@ public virtual int GetBuildTime(ActorInfo unit, BuildableInfo bi)
+				time = GetProductionCost(unit);
+
+			var modifiers = unit.TraitInfos<IProductionTimeModifierInfo>().Select(t => t.GetProductionTimeModifier(techTree, Info.Type));
+			time = Util.ApplyPercentageModifiers(time, modifiers);
```
This can simplify to
```csharp
var modifiers = unit.TraitInfos<IProductionTimeModifierInfo>()
	.Select(t => t.GetProductionTimeModifier(techTree, Info.Type))
	.Append(bi.BuildDurationModifier)
	.Append(Info.BuildDurationModifier);

return Util.ApplyPercentageModifiers(time, modifiers);
```

pchote commented on OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:
```diff
@@ -434,15 +437,28 @@ public virtual int GetBuildTime(ActorInfo unit, BuildableInfo bi)
+			if (valued == null)
+				return 0;
+
+			var modifiers = unit.TraitInfos<IProductionCostModifierInfo>().Select(t => t.GetProductionCostModifier(techTree, Info.Type));
```
Nit: it helps readability if the `.Select` is put on its own line.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/ProductionTooltipLogic.cs:
```diff
@@ -70,8 +70,12 @@ public ProductionTooltipLogic(Widget widget, TooltipContainerWidget tooltipConta
+				var cost = 0;
+				if (tooltipIcon.ProductionQueue != null)
+					cost = tooltipIcon.ProductionQueue.GetProductionCost(actor);
+				else if (actor.TraitInfoOrDefault<ValuedInfo>() != null)
```
How can we have a tooltip if we don't have a production queue?

If we do need this fallback, please reduce it to a single query using
```csharp
else
{
	var valued = actor.TraitInfoOrDefault<ValuedInfo>();
	if (valued != null)
		cost = valued.Cost;
}

MustaphaTR commented:
Updated.

pchote commented:
This now crashes immediately on game start, because the `ProductionQueue` constructor calls `CacheProducibles` before `techTree` has been assigned.

Moving the `CacheProducibles` call to `Created` fixes the issue.  The `playerResources` and `developerMode` should move there too for consistency.

LGTM after this has been fixed.

MustaphaTR commented:
Updated.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8d18f310c1bf0419a8b15545beb75965d3de6c58...9c5ae5a0989fca9f72d4e6113713c9f8278e35d4
